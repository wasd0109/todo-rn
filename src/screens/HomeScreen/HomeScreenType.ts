import { StackNavigationProp } from "@react-navigation/stack";
import { RootStackParamList } from "../../types/ReactNavigationType";

type HomeScreenNavigationProp = StackNavigationProp<RootStackParamList, "Home">;

export type HomeScreenProps = {
  navigation: HomeScreenNavigationProp;
};
