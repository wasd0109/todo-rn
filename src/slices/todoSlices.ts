import { createSlice, PayloadAction } from "@reduxjs/toolkit";

export interface Todo {
  title: string;
  date: number;
  id: string;
}

export interface TodoState {
  todoList: Todo[];
}

const initialState: TodoState = {
  todoList: [],
};

export const todoSlice = createSlice({
  name: "todo",
  initialState,
  reducers: {
    addTodo: (state, action: PayloadAction<Todo>) => {
      state.todoList.push(action.payload);
    },
  },
});

export const { addTodo } = todoSlice.actions;

export default todoSlice.reducer;
