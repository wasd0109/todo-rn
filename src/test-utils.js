import React from "react";
import "react-native-paper";
import { render } from "@testing-library/react-native";
import { Provider as PaperProvider } from "react-native-paper";
import { NavigationContainer } from "@react-navigation/native";
import { createStackNavigator } from "@react-navigation/stack";

const AllTheProviders = ({ children }) => {
  return <PaperProvider>{children}</PaperProvider>;
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from "@testing-library/react-native";

// override render method
export { customRender as render };

const Stack = createStackNavigator();
const MockedNavigator = ({ screenName, component, params = {} }) => {
  return (
    <NavigationContainer>
      <Stack.Navigator>
        <Stack.Screen
          name={screenName}
          component={component}
          initialParams={params}
        />
      </Stack.Navigator>
    </NavigationContainer>
  );
};

export default MockedNavigator;
