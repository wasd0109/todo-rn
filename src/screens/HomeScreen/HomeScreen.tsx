import React, { useState } from "react";
import { StyleSheet, Text, View, FlatList } from "react-native";
import { List, Divider, Provider, IconButton } from "react-native-paper";
import { useSelector } from "react-redux";
import { RootState } from "../../store";
import { HomeScreenProps } from "./HomeScreenType";

const HomeScreen = ({ navigation }: HomeScreenProps) => {
  const todoList = useSelector((state: RootState) => state.todo.todoList);
  return (
    <FlatList
      data={todoList}
      keyExtractor={(item) => item.id}
      renderItem={({ item }) => {
        return (
          <>
            <List.Item
              title={item.title}
              right={(props) => (
                <IconButton
                  {...props}
                  icon="dots-horizontal"
                  onPress={() => navigation.navigate("Detail", { id: item.id })}
                  accessibilityLabel="more-button"
                />
              )}
            />
            <Divider />
          </>
        );
      }}
    />
  );
};

//

export default HomeScreen;

const styles = StyleSheet.create({});
