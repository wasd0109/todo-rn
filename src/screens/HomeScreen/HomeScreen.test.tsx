import React from "react";
import { render } from "../../test-utils";
import HomeScreen from ".";
import { fireEvent, RenderAPI } from "@testing-library/react-native";
import { Provider } from "react-redux";
import { configureStore } from "@reduxjs/toolkit";
import { todoSlice, addTodo } from "../../slices/todoSlices";

let component: RenderAPI;

let store;

const initialTodo = {
  title: "Test",
  date: 123456,
  id: "1f4ba8ae-f657-421e-8bb7-48e39f3be323",
};

const props = {
  navigation: {
    navigate: jest.fn(),
  },
} as any;

beforeEach(() => {
  store = configureStore({ reducer: { todo: todoSlice.reducer } });
  store.dispatch(addTodo(initialTodo));
  component = render(
    <Provider store={store}>
      <HomeScreen {...props} />
    </Provider>
  );
});

describe("HomeScreen render correctly", () => {
  test("HomerScreen render", () => {
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("HomeScreen render list correctly", () => {
    const { getAllByText } = component;
    expect(getAllByText(initialTodo.title).length).toEqual(1);
  });
});

describe("HomeScreen function properly", () => {
  test("Detail button lead to detail screen with correct parameters", () => {
    const { getByLabelText } = component;
    const detailBtnRef = getByLabelText("more-button");
    fireEvent.press(detailBtnRef);
    expect(props.navigation.navigate).toBeCalled();
    expect(props.navigation.navigate).toBeCalledWith("Detail", {
      id: initialTodo.id,
    });
  });
});
