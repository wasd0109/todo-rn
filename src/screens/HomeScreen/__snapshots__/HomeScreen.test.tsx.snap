// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen render correctly HomerScreen render 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    data={
      Array [
        Object {
          "date": 123456,
          "id": "1f4ba8ae-f657-421e-8bb7-48e39f3be323",
          "title": "Test",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          focusable={false}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              false,
              Array [
                Object {
                  "padding": 8,
                },
                undefined,
              ],
            ]
          }
        >
          <View
            style={
              Object {
                "flexDirection": "row",
              }
            }
          >
            <View
              style={
                Array [
                  Object {
                    "marginVertical": 6,
                    "paddingLeft": 8,
                  },
                  Object {
                    "flex": 1,
                    "justifyContent": "center",
                  },
                ]
              }
            >
              <Text
                numberOfLines={1}
                selectable={false}
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontWeight": "400",
                    },
                    Object {
                      "textAlign": "left",
                    },
                    Array [
                      Object {
                        "fontSize": 16,
                      },
                      Object {
                        "color": "rgba(0, 0, 0, 0.87)",
                      },
                      undefined,
                    ],
                  ]
                }
              >
                Test
              </Text>
            </View>
            <View
              accessibilityLabel="more-button"
              accessibilityRole="button"
              accessibilityState={
                Object {
                  "disabled": undefined,
                }
              }
              accessible={true}
              focusable={true}
              hitSlop={
                Object {
                  "bottom": 6,
                  "left": 6,
                  "right": 6,
                  "top": 6,
                }
              }
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Array [
                  Object {
                    "overflow": "hidden",
                  },
                  Array [
                    Object {
                      "alignItems": "center",
                      "justifyContent": "center",
                      "margin": 6,
                      "overflow": "hidden",
                    },
                    Object {
                      "borderRadius": 18,
                      "height": 36,
                      "width": 36,
                    },
                    undefined,
                    Object {
                      "marginRight": 0,
                      "marginVertical": 0,
                    },
                  ],
                ]
              }
            >
              <View>
                <Text />
              </View>
            </View>
          </View>
        </View>
        <View
          style={
            Array [
              Object {
                "backgroundColor": "rgba(0, 0, 0, 0.12)",
                "height": 0.5,
              },
              undefined,
              undefined,
            ]
          }
        />
      </View>
    </View>
  </RCTScrollView>
</View>
`;
